enhancements/network/user-defined-network-segmentation.md

```diff
@@ -413,6 +519,203 @@ guarantee uniqueness and eliminates the ability to falsify access.
+by the contoller as follows:
+1. Check the owner reference match the request CR UID.
+2. OVN-K user-defined-network finalizer exist.
+3. NAD spec correspond to desired spec.
+In case one of the following checks fails, the controller should reconcile the request and 
+this state in the CR status network is not ready.
+
+In scenario of cluster-scope CRD request, the controller should do best effort to create the desired NAD in each
+specified namespace.
+In case on or more NAD creation fails (e.g.: due to namespace not exist), the controller should continue to the 
+next namespace.
+When finished, reflect in the status all namespaces where NAD creation failed to create.
+
+##### General flow
+1. On namespaced-scope CRD creation:
+  1.1. Validates the CRD spec. 
+  1.2. Generate NAD manifest from the CRD spec.
+  1.3. Check no the desired NAD is not created yet, if so, create the NAD and return.
+  1.4. Otherwise, verify the existing NAD correspond to desired spec, if so return.
+  1.5. In case the NAD is malformed raise an error and return.
+  1.6. Update the status as follows:
+   1.6.1. Reflect namespaces where network is available.
+   1.6.2. Reflect the reconciliation errors, which namespaces failed and the reason.
+
+2. on cluster-scope CRD creation:
+   2.1. Validates the CRD spec.
+   2.2. Generate NAD manifest from the CRD spec.
+   2.3. For each namespace specified in the spec:
+       2.3.1. Check no the desired NAD is not created yet, if so, create the NAD and continue. 
+       2.3.2. Otherwise, verify the existing NAD correspond to desired spec, if so continue.
+       2.3.3. In case the NAD is malformed raise an error and continue.
+   2.4. Update the status as follows:
+       2.4.1. Reflect namespaces where network is available.
+       2.4.1. Reflect the reconciliation errors, which namespaces failed and the reason.
+
+3. On namespace-scope CRD deletion:
+   3.1. If no NAD exist, return.
+   3.1. In case a NAD exist, ensure no pod specifying the network.
+     3.1.1. In case no pod specifying the network, remove the finalizer, allowing the cluster garbage collector dispose the NAD object.
+     3.1.2. Otherwise, reflect in the status the network is being deleted because its in use.
+
+4. On cluster-scope CRD deletion:
+  4.1. For each namespace specified in the spec:
+   4.1. In case a NAD exist, ensure no pod specifying the network.
+   4.1.1. In case no pod specifying the network, remove the finalizer, allowing the cluster garbage collector dispose the NAD object.
+   4.1.2. Otherwise, reflect in the status the network is cannot be deleted because its in use.
+
+##### NetworkAttachmentDefinition rendering
+The underlying OVN network ID (network-name) is represented by the NAD CNI's network config `name` field, and must be unique.
+
+The network-name is generated by the controller and should not be exposed for modification.
```

Comment:
> > really ? wouldn't it be simpler to just expose this to the user ?
> 
> I assume you meant for the cluster-scope CRD case. Exposing the name may introduce issues in scenario there are two CRD instances specifying the same network-name. Having multiple CRDs pointing to the same network-name is kind of missing the point, its error prune and may get confusing to manage. I think the relation between the CRD and underlaying network (represented by the netwrok name) should be 1:1.

It should be 1:1. What in my proposal (especially the one where I say to use the CRD metadata name as the network name) breaks this 1:1 association ?

> 
> > Or we could simply grab the `ClusterUserDefinedNetwork.Metadata.Name` attribute, couldn't we ? That seems more kubernetes native.
> 
> I would rather have this option over exposing the network-name. There was a discussion about this option while ago.

Would you point to it ? Right now I see no reason why this is preferable. 

> 
> I think its safer to use UUIDs, it could prevents for one to guess the network-name and tap in.

The only person that can "tap in" here is the cluster admin right ? What's wrong with that ? 

I guess I just don't understand the advantages of using a UUID for a network name.

Reply:
My examples here show using the name:

https://github.com/openshift/enhancements/pull/1638#issuecomment-2195722725

I think the above method is sound and is more clear than using UUIDs.